processing/src/main/java/org/apache/druid/segment/join/filter/rewrite/JoinFilterPreAnalysisGroup.java

```diff
@@ -0,0 +1,185 @@
+ */
+
+package org.apache.druid.segment.join.filter.rewrite;
+
+import org.apache.druid.query.filter.Filter;
+import org.apache.druid.segment.VirtualColumns;
+import org.apache.druid.segment.join.JoinableClause;
+import org.apache.druid.segment.join.filter.JoinFilterAnalyzer;
+import org.apache.druid.segment.join.filter.JoinFilterPreAnalysis;
+import org.apache.druid.segment.join.filter.JoinableClauses;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A JoinFilterPreAnalysisGroup holds all of the JoinFilterPreAnalysis objects for a given query and
+ * also stores the per-query parameters that control the filter rewrite operations (from the query context).
+ *
+ * The analyses map is keyed by (Filter, JoinableClause list, VirtualColumns): each Filter in the map belongs to a
+ * separate level of query (e.g. outer query, subquery level 1, etc.)
+ *
+ * A concurrent hash map is used since the per-level Filter processing occurs across multiple threads.
+ */
+public class JoinFilterPreAnalysisGroup
+{
+  private final JoinFilterRewriteConfig joinFilterRewriteConfig;
+  private final ConcurrentHashMap<JoinFilterPreAnalysisGroupKey, JoinFilterPreAnalysis> analyses;
+
+  /**
+   * This is an undocumented option provided as a transition tool:
+   *
+   * The join filter rewrites originally performed the pre-analysis phase prior to any per-segment processing,
+   * analyzing only the filter in the top-level of the query.
+   *
+   * This did not work for nested queries (see https://github.com/apache/druid/pull/9978), so the rewrite pre-analysis
+   * was moved into the cursor creation of the {@link org.apache.druid.segment.join.HashJoinSegmentStorageAdapter}.
+   * This design requires synchronization across multiple segment processing threads; the old rewrite mode
+   * is kept temporarily available in case issues arise with the new mode, and the user does not run queries with the
+   * affected nested shape.
+   */
+  private JoinFilterPreAnalysis preAnalysisForOldRewriteMode;
+
+  public JoinFilterPreAnalysisGroup(
+      JoinFilterRewriteConfig joinFilterRewriteConfig
+  )
+  {
+    this.joinFilterRewriteConfig = joinFilterRewriteConfig;
+    this.analyses = new ConcurrentHashMap<>();
+  }
+
+  public JoinFilterRewriteConfig getJoinFilterRewriteConfig()
+  {
+    return joinFilterRewriteConfig;
+  }
+
+  public JoinFilterPreAnalysis computeJoinFilterPreAnalysisIfAbsent(
+      Filter filter,
+      List<JoinableClause> clauses,
+      VirtualColumns virtualColumns
+  )
+  {
+    // Some filters have potentially expensive hash codes that are lazily computed and cached.
+    // We call hashCode() here in a synchronized block before we attempt to use the Filter in the analyses map,
+    // to ensure that the hashCode is only computed once per Filter since the Filter interface is not thread-safe.
+    synchronized (analyses) {
+      if (filter != null) {
+        filter.hashCode();
```

Comment:
Oh, is this why we don't need thread-safety on filter hashcode methods i guess? This seems kind of a funny way to prime them with the cached values, I think maybe the supplier.memoize pattern would be a little cleaner and make this not necessary?

Reply:
Removed this after changing to supplier.memoize